src/ert/ensemble_evaluator/evaluator_async.py

```diff
@@ -0,0 +1,471 @@
+            await self._signal_cancel()  # let ensemble know it should stop
+
+    @property
+    def config(self) -> EvaluatorServerConfig:
+        return self._config
+
+    @property
+    def ensemble(self) -> Ensemble:
+        return self._ensemble
+
+    async def _create_cloud_event(
+        self,
+        event_type: str,
+        data: Optional[Dict[str, Any]] = None,
+        extra_attrs: Optional[Dict[str, Any]] = None,
+    ) -> CloudEvent:
+        """Returns a CloudEvent with the given properties"""
+        if isinstance(data, dict):
+            data["iter"] = self._iter
+        if extra_attrs is None:
+            extra_attrs = {}
+
+        attrs = {
+            "type": event_type,
+            "source": f"/ert/ensemble/{self.ensemble.id_}",
+        }
+        attrs.update(extra_attrs)
+        return CloudEvent(
+            attrs,
+            data,
+        )
+
+    async def _create_cloud_message(
+        self,
+        event_type: str,
+        data: Optional[Dict[str, Any]] = None,
+        extra_attrs: Optional[Dict[str, Any]] = None,
+        data_marshaller: Optional[Callable[[Any], Any]] = evaluator_marshaller,
+    ) -> str:
+        """Creates the CloudEvent and returns the serialized json-string"""
+        event = await self._create_cloud_event(event_type, data, extra_attrs)
+        return to_json(event, data_marshaller=data_marshaller).decode()
+
+    @contextmanager
+    def store_client(
+        self, websocket: WebSocketServerProtocol
+    ) -> Generator[None, None, None]:
+        self._clients.add(websocket)
+        yield
+        self._clients.remove(websocket)
+
+    async def handle_client(
+        self, websocket: WebSocketServerProtocol, path: str
+    ) -> None:
+        with self.store_client(websocket):
+            current_snapshot_dict = self._ensemble.snapshot.to_dict()
+            event = await self._create_cloud_message(
+                EVTYPE_EE_SNAPSHOT, current_snapshot_dict
+            )
+            await websocket.send(event)
+
+            async for message in websocket:
+                client_event = from_json(
+                    message, data_unmarshaller=evaluator_unmarshaller
+                )
+                logger.debug(f"got message from client: {client_event}")
+                if client_event["type"] == EVTYPE_EE_USER_CANCEL:
+                    logger.debug(f"Client {websocket.remote_address} asked to cancel.")
+                    await self._signal_cancel()
+
+                elif client_event["type"] == EVTYPE_EE_USER_DONE:
+                    logger.debug(f"Client {websocket.remote_address} signalled done.")
+                    await self._stop()
+
+    @asynccontextmanager
+    async def count_dispatcher(self) -> AsyncIterator[None]:
+        await self._dispatchers_connected.put(None)
+        yield
+        await self._dispatchers_connected.get()
+        self._dispatchers_connected.task_done()
```

Comment:
It seems we are only putting an object in the queue here, and then getting it after exiting context. Can we use a Future or Event instead, and check if it is set rather than checking if the queue is empty?